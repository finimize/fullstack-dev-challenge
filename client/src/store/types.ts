export const UPDATE_DETAILS = 'UPDATE_DETAILS'
export const UPDATE_CALCULATIONS = 'UPDATE_CALCULATIONS'
export const UPDATE_INTEREST = 'UPDATE_INTEREST'
export const VALIDATE_DETAILS = 'VALIDATE_DETAILS'
export const NEXT_PAGE = 'NEXT_PAGE'
export const PREVIOUS_PAGE = 'PREVIOUS_PAGE'
export const GOT_TO_PAGE = 'GOT_TO_PAGE'
export const TOGGLE_MODE = 'TOGGLE_MODE'
export const TOGGLE_COMPOUNDING_FREQUENCY = 'TOGGLE_COMPOUNDING_FREQUENCY'
export const UPDATE_DATA = 'UPDATE_DATA'
