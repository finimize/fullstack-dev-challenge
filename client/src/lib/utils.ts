import type { ChartTooltipItem } from 'chart.js'
import { SAVINGS_DEFAULTS } from './constants'

export const createQueryString = (queries: { [key: string]: any }): string =>
    Object.keys(queries)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(queries[key]))
        .join('&')

export const appendYearsQuery = (queryString: string) =>
    `${queryString}&years=${SAVINGS_DEFAULTS.YEARS}`

export const extractYearlyData = (projections: number[]) =>
    projections.filter((item, index) => index % 12 === 0)

export const numberFormatter = new Intl.NumberFormat('gb-GB', {
    style: 'currency',
    currency: 'GBP',
})

export const getTooltipTitle = (tooltipItems: ChartTooltipItem[]) => `Year ${tooltipItems[0].label}`

export const getTooltipLabel = (tooltipItem: ChartTooltipItem) => {
    const value = tooltipItem.value ? parseFloat(tooltipItem.value) : 0
    return numberFormatter.format(value)
}
