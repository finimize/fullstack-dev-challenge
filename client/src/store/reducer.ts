import {
    UPDATE_DETAILS,
    UPDATE_CALCULATIONS,
    UPDATE_INTEREST,
    UPDATE_DATA,
    VALIDATE_DETAILS,
    NEXT_PAGE,
    TOGGLE_MODE,
    TOGGLE_COMPOUNDING_FREQUENCY,
    PREVIOUS_PAGE,
    GOT_TO_PAGE,
} from './types'
import {
    StateInterface,
    UpdateDetailsTypeInterface,
    UpdateCalculationsTypeInterface,
    UpdateInterestRateInterface,
    UpdateDataInterface,
    ValidateDetailsTypeInterface,
    NextPageTypeInterface,
    ToggleModeInterface,
    ToggleCompoundingFrequencyInterface,
    PreviousPageTypeInterface,
    CompoundingFrequency,
    GoToPageInterface,
} from './store.interface'

export const reducer = (
    initialState: StateInterface,
    action:
        | UpdateDetailsTypeInterface
        | UpdateCalculationsTypeInterface
        | UpdateInterestRateInterface
        | UpdateDataInterface
        | ValidateDetailsTypeInterface
        | NextPageTypeInterface
        | PreviousPageTypeInterface
        | ToggleModeInterface
        | ToggleCompoundingFrequencyInterface
        | GoToPageInterface,
): StateInterface => {
    switch (action.type) {
        case NEXT_PAGE: {
            return {
                ...initialState,
                currentPage: initialState.currentPage + 1,
            }
        }
        case GOT_TO_PAGE: {
            return {
                ...initialState,
                currentPage: action.payload,
            }
        }
        case PREVIOUS_PAGE: {
            return {
                ...initialState,
                currentPage: initialState.currentPage - 1,
            }
        }
        case TOGGLE_MODE: {
            return {
                ...initialState,
                calculatorMode: action.payload,
            }
        }
        case TOGGLE_COMPOUNDING_FREQUENCY: {
            return {
                ...initialState,
                calculations: {
                    ...initialState.calculations,
                    compoundingFrequency:
                        CompoundingFrequency[action.payload as keyof typeof CompoundingFrequency],
                },
            }
        }
        case UPDATE_DETAILS:
            return {
                ...initialState,
                details: {
                    ...initialState.details,
                    [action.field]: action.payload,
                },
            }
        case UPDATE_CALCULATIONS:
            return {
                ...initialState,
                calculations: {
                    ...initialState.calculations,
                    [action.field]: action.payload,
                },
            }
        case UPDATE_INTEREST:
            return {
                ...initialState,
                calculations: {
                    ...initialState.calculations,
                    interestRate: action.payload,
                },
            }
        case UPDATE_DATA:
            return {
                ...initialState,
                data: {
                    ...initialState.data,
                    ...action.payload,
                },
            }
        case VALIDATE_DETAILS:
            return {
                ...initialState,
                details: {
                    ...initialState.details,
                    firstName: {
                        ...initialState.details.firstName,
                        error: action.payload.firstName,
                    },
                    lastName: {
                        ...initialState.details.lastName,
                        error: action.payload.lastName,
                    },
                    email: {
                        ...initialState.details.email,
                        error: action.payload.email,
                    },
                },
            }
        /* istanbul ignore next */
        default:
            return initialState
    }
}
