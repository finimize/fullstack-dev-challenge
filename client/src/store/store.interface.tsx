import {
    UPDATE_DETAILS,
    UPDATE_CALCULATIONS,
    UPDATE_INTEREST,
    UPDATE_DATA,
    VALIDATE_DETAILS,
    NEXT_PAGE,
    TOGGLE_MODE,
    TOGGLE_COMPOUNDING_FREQUENCY,
    PREVIOUS_PAGE,
} from './types'

export interface TextInputInterface {
    value: string
    error: string
}

interface DetailsInterface {
    firstName: TextInputInterface
    lastName: TextInputInterface
    email: TextInputInterface
}

export interface DataValueInterface {
    yearlySavings: number[]
    finalValue: number
    yearlyBreakdown: {
        yearFinal: number
        savings: []
    }[]
}

export interface DataInterface {
    value: DataValueInterface | null
    error: string
    isLoading: boolean
}

export interface CalculationsInterface {
    initialSavings: TextInputInterface
    interestRate: number
    compoundingFrequency: number
    monthlyContributions: TextInputInterface
}

export interface StateInterface {
    details: DetailsInterface
    data: DataInterface
    currentPage: number
    calculatorMode: string
    calculations: CalculationsInterface
}

export type UpdateDetailsTypeInterface = {
    type: typeof UPDATE_DETAILS
    field: 'firstName' | 'lastName' | 'email'
    payload: Partial<TextInputInterface>
}
export type UpdateCalculationsTypeInterface = {
    type: typeof UPDATE_CALCULATIONS
    field: 'initialSavings' | 'monthlyContributions'
    payload: Partial<TextInputInterface>
}

export type UpdateInterestRateInterface = {
    type: typeof UPDATE_INTEREST
    payload: number
}

export type AppProps = {
    children: React.ReactNode
}

export type ValidateDetailsTypeInterface = {
    type: typeof VALIDATE_DETAILS
    payload: {
        firstName: string
        lastName: string
        email: string
    }
}

export type NextPageTypeInterface = {
    type: typeof NEXT_PAGE
}

export type PreviousPageTypeInterface = {
    type: typeof PREVIOUS_PAGE
}

export type ToggleModeInterface = {
    type: typeof TOGGLE_MODE
    payload: string
}

export enum CompoundingFrequency {
    annually = 1,
    quarterly = 4,
    monthly = 12,
}

export type ToggleCompoundingFrequencyInterface = {
    type: typeof TOGGLE_COMPOUNDING_FREQUENCY
    payload: string
}

export type UpdateDataInterface = {
    type: typeof UPDATE_DATA
    payload: DataInterface
}
