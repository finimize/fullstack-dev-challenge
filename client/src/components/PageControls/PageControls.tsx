import React, { FC, useContext } from 'react'
import { Box, Button } from '@chakra-ui/react'
import { AppContext } from '../../store'

interface PageControlProps {
    prevPage?: () => void
    nextPage?: () => void
}

export const PageControls: FC<PageControlProps> = ({ prevPage, nextPage }) => {
    const { state } = useContext(AppContext)
    const buttonJustification: () => string = () => {
        if (state.currentPage === 1) return 'flex-end'
        if (state.currentPage === 3) return 'flex-start'
        return 'space-between'
    }

    return (
        <Box justifyContent={buttonJustification()} display='flex' padding='8'>
            {state.currentPage !== 1 && (
                <Button
                    data-testid='previous-button'
                    backgroundColor='primary'
                    width='24'
                    _hover={{
                        bg: 'blue400',
                    }}
                    color='white'
                    boxShadow='-1px 1px 3px 0px #a7a7a7'
                    onClick={prevPage}
                >
                    Back
                </Button>
            )}
            {state.currentPage !== 3 && (
                <Button
                    data-testid='next-button'
                    width='24'
                    borderColor='blue400'
                    borderWidth='2px'
                    backgroundColor='primary'
                    boxShadow='-1px 1px 3px 0px #a7a7a7'
                    color='white'
                    _hover={{
                        bg: 'blue400',
                    }}
                    onClick={nextPage}
                >
                    Next
                </Button>
            )}
        </Box>
    )
}
