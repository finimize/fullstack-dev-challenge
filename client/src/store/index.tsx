import React, { createContext, useReducer, FC } from 'react'
import { initialState } from './initialState'
import { reducer } from './reducer'
import {
    UpdateDetailsTypeInterface,
    ValidateDetailsTypeInterface,
    NextPageTypeInterface,
    StateInterface,
    AppProps,
    PreviousPageTypeInterface,
    ToggleModeInterface,
    UpdateCalculationsTypeInterface,
    UpdateInterestRateInterface,
} from './store.interface'

const AppContext = createContext<{
    state: StateInterface
    dispatch: React.Dispatch<
        | UpdateDetailsTypeInterface
        | UpdateCalculationsTypeInterface
        | ValidateDetailsTypeInterface
        | NextPageTypeInterface
        | ToggleModeInterface
        | PreviousPageTypeInterface
        | UpdateInterestRateInterface
    >
}>({
    state: initialState,
    dispatch: () => null,
})

const AppProvider: FC<AppProps> = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState)
    return <AppContext.Provider value={{ state, dispatch }}>{children}</AppContext.Provider>
}

export { AppContext, AppProvider }
