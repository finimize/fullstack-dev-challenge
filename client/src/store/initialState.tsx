import { StateInterface, CompoundingFrequency } from './store.interface'

export const initialState: StateInterface = {
    details: {
        firstName: {
            value: '',
            error: '',
        },
        lastName: {
            value: '',
            error: '',
        },
        email: {
            value: '',
            error: '',
        },
    },
    data: {
        value: null,
        error: '',
        isLoading: false,
    },
    calculations: {
        initialSavings: {
            value: '10',
            error: '',
        },
        interestRate: 20,
        compoundingFrequency: CompoundingFrequency.annually,
        monthlyContributions: {
            value: '0',
            error: '',
        },
    },
    calculatorMode: 'Simple',
    currentPage: 1,
}
