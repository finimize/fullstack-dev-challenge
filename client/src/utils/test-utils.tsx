import React, { FC, useReducer } from 'react'
import { ChakraProvider, extendTheme } from '@chakra-ui/react'
import { render as rtlRender, RenderResult, RenderOptions } from '@testing-library/react'
import { AppContext, DispatchType } from '../store'
import { initialState } from '../store/initialState'
import { reducer } from '../store/reducer'
import { StateInterface } from '../store/store.interface'
import { theme } from '../theme'

const mainTheme = extendTheme(theme)

interface CustomRenderOptions extends Partial<RenderOptions> {
    store?: StateInterface
    defaultTheme?: typeof theme
}

export interface CustomRenderInterface {
    (
        ui: React.ReactElement<unknown, string | React.JSXElementConstructor<unknown>>,
        options: CustomRenderOptions,
    ): { render: RenderResult; dispatchMock: jest.Mock<unknown, unknown[]> }
}

const customRender: CustomRenderInterface = (
    ui,
    { store = initialState, defaultTheme = mainTheme, ...options },
) => {
    const dispatchMock = jest.fn()
    const AllProviders: FC = ({ children }) => {
        // const [calculatorState, setCalculatorState] = useState(store)
        const [calculatorState, dispatch] = useReducer(reducer, store)
        return (
            <ChakraProvider theme={defaultTheme}>
                <AppContext.Provider
                    value={{
                        state: calculatorState,
                        dispatch: dispatchMock.mockImplementation((value: DispatchType) =>
                            dispatch(value),
                        ),
                    }}
                >
                    {' '}
                    {children}{' '}
                </AppContext.Provider>{' '}
            </ChakraProvider>
        )
    }
    return { render: rtlRender(ui, { wrapper: AllProviders, ...options }), dispatchMock }
}

export { customRender }
