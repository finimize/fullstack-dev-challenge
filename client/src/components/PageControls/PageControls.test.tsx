import React from 'react'
import { cleanup } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { initialState } from '../../store/initialState'
import { customRender } from '../../utils/test-utils'
import { PageControls } from './PageControls'

afterEach(cleanup)

describe('component/PageControls', () => {
    it('allows to both buttons to be pressed on second page', () => {
        const nextMock = jest.fn()
        const prevMock = jest.fn()
        const { render } = customRender(<PageControls nextPage={nextMock} prevPage={prevMock} />, {
            store: { ...initialState, currentPage: 2 },
        })
        const { getByTestId, container } = render
        expect(container).toMatchSnapshot()
        const prevbutton = getByTestId('previous-button')
        const nextbutton = getByTestId('next-button')
        userEvent.click(prevbutton)
        expect(prevMock).toHaveBeenCalled()
        userEvent.click(nextbutton)
        expect(nextMock).toHaveBeenCalled()
    })

    it('allows only next button on 1st page', () => {
        const nextMock = jest.fn()
        const prevMock = jest.fn()
        const { render } = customRender(
            <PageControls nextPage={nextMock} prevPage={prevMock} />,
            {},
        )
        const { getByTestId, queryByTestId, container } = render
        expect(container).toMatchSnapshot()
        // Query the previous button as it should not be in the document
        const prevbutton = queryByTestId('previous-button')
        const nextbutton = getByTestId('next-button')
        expect(prevbutton).not.toBeInTheDocument()
        userEvent.click(nextbutton)
        expect(nextMock).toHaveBeenCalled()
    })

    it('allows only previous button on 3st page', () => {
        const nextMock = jest.fn()
        const prevMock = jest.fn()
        const { render } = customRender(<PageControls nextPage={nextMock} prevPage={prevMock} />, {
            store: { ...initialState, currentPage: 3 },
        })
        const { getByTestId, queryByTestId, container } = render
        // Query the previous button as it should not be in the document
        expect(container).toMatchSnapshot()
        const prevbutton = getByTestId('previous-button')
        const nextbutton = queryByTestId('next-button')
        expect(nextbutton).not.toBeInTheDocument()
        userEvent.click(prevbutton)
        expect(prevMock).toHaveBeenCalled()
    })
})
