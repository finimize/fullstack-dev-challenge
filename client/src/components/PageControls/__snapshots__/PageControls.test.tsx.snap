// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/PageControls allows only next button on 1st page 1`] = `
<div>
   
  <div
    class="css-j4xgao"
  >
    <button
      class="chakra-button css-1mi4eb0"
      data-testid="next-button"
      type="button"
    >
      Next
    </button>
  </div>
   
   
</div>
`;

exports[`component/PageControls allows only previous button on 3st page 1`] = `
<div>
   
  <div
    class="css-fo25ua"
  >
    <button
      class="chakra-button css-a7myp6"
      data-testid="previous-button"
      type="button"
    >
      Back
    </button>
  </div>
   
   
</div>
`;

exports[`component/PageControls allows to both buttons to be pressed on second page 1`] = `
<div>
   
  <div
    class="css-jo7bsi"
  >
    <button
      class="chakra-button css-a7myp6"
      data-testid="previous-button"
      type="button"
    >
      Back
    </button>
    <button
      class="chakra-button css-1mi4eb0"
      data-testid="next-button"
      type="button"
    >
      Next
    </button>
  </div>
   
   
</div>
`;
